### PR TITLE
global sched/pool.  Task objects for ease of async use.

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

File: synapse/lib/task.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        self.info = {}
+        self.iden = iden
+
+        self.on('task:fini', self._onTaskFini)
+
+    def _onTaskFini(self, mesg):
+
+        retn = mesg[1].get('retn')
+        if retn is not None:
+            self.fire('task:retn', retn=retn)
```
Comment (vertexmc) [inline synapse/lib/task.py]: should this message include the iden?

Comment (invisig0th) [reply synapse/lib/task.py]: yep!  good catch!

File: synapse/lib/task.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    def _task_run(self):
+        raise s_common.NoSuchImpl(name='_task_run')
+
+    def __call__(self):
+        self.run()
+
+class CallTask(Task):
+
+    '''
+    An extnsion for a runnable task.
```
Comment (vEpiphyte) [inline synapse/lib/task.py]: ``extension``

File: synapse/glob.py [hunk 3]
```diff
@@ -2,14 +2,8 @@
...
 
-# A global Plex instance for managing socket i/o.
-plex = None
-
-# A global Sched instance for maint routines
-sched = None
-
-# A host info dictionary for "thishost" API
-hostinfo = None
+pool = s_threads.Pool()
```
Comment (vEpiphyte) [inline synapse/glob.py]: The pool ``__ini__`` sig is the following:
``
def __init__(self, size=3, maxsize=None):
``

This results in a pool of global pool of three threads max to run things from.

Comment (vEpiphyte) [reply synapse/glob.py]: This seems like a small pool for a global thread pool

Comment (vEpiphyte) [reply synapse/glob.py]: For context, the ThreadPoolExecutor from concurrent.futures uses the following for the default number of max workers ``max_workers = (os.cpu_count() or 1) * 5``

File: synapse/lib/threads.py [hunk 4]
```diff
@@ -226,26 +247,13 @@ def _run_work(self):
...
+                func, args, kwargs = work
+                func(*args, **kwargs)
 
             except Exception as e:
-
-                if jid is not None:
-                    self.fire('job:done', jid=jid, **s_common.excinfo(e))
-
-            self.fire('pool:work:fini', work=work)
+                logger.exception(e)
```
Comment (vEpiphyte) [inline synapse/lib/threads.py]: ``logger.exception('Error while executing task')`` is all that is needed here.  logger.exception pulls the exception data from the stack directly.

File: synapse/glob.py (hunk 3 above)
Comment (invisig0th) [reply synapse/glob.py]: Actually, it ends up with a thread pool of *minimum* size 3, with no maximum size, which will keep >size threads alive until there is not more work to do, but then let them die off ( to a minimum of size, so the next requests that come in dont have to fire a new one ).  I was considering setting a max, but most of our use cases so far haven't used one.

File: synapse/lib/task.py (hunk 2 above)
Comment (invisig0th) [reply synapse/lib/task.py]: haha. thx.

File: synapse/lib/threads.py (hunk 4 above)
Comment (invisig0th) [reply synapse/lib/threads.py]: ah, yea.  good call.  dunno where i got into the paradigm of logger.exception(Exception) but you're def right.

**Pull request merged**